File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        # base case: check if root is empty
        # if root is empty then create new TreeNode
        if not root:
            return TreeNode(val)

        # case 1: val > root.val
        if val > root.val:
            root.right = self.insertIntoBST(root.right, val)

        # case 2: val < root.val
        if val < root.val:
            root.left = self.insertIntoBST(root.left, val)

        return root
    

        
        

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int totalCount = 0;
    public int goodNodes(TreeNode root) {
        if(root == null) return 0;
        dfs(root, root.val);
        return totalCount;
    }

    public void dfs(TreeNode node, int maxSoFar) {
        if(node == null) return;

        if(node.val >= maxSoFar) {
            totalCount++;
        }

        maxSoFar = Math.max(maxSoFar, node.val);

        dfs(node.left, maxSoFar);
        dfs(node.right, maxSoFar);
    }
}
